designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,505 @@
+1. Users will have to know which version of ESLint supports `extends` in order to use it. There really isn't an easy way to feature test this capability other than to inspect the version of the current ESLint package.
+
+## Backwards Compatibility Analysis
+
+This proposal is additive and does not affect the way existing configurations are evaluated.
+
+## Alternatives
+
+1. **A utility package.** Instead of including `extends` in flat config itself, we could create a utility package that provides a method to accomplish the same thing. This has the advantage that ESLint v8 users could also use this utility package, but the downside that it would require users to install Yet Another Package to do something that many feel should be available out of the box.
+1. **A utility function.** A variation of the previous approach is to export a function from the `eslint` package that people could use to extend configs. This has the advantage that it would be easier to feature test for but the downside that it still requires an extra step to use `extends`.
+
+## Open Questions
+
+1. **Does the `#` pattern in plugin configs cause any unintended side effects?** The string `"#"` would be an invalid package name, so it's unlikely to cause any naming collisions.
+1. **Should `extends` be allowed in plugin configs?** This RFC does not allow `extends` in plugin configs for simplicity. Allowing this would mean the possibility of nested `extends`, which may be desirable but also more challenging to implement. Further, if plugins export configs with `extends`, then that automatically means those configs cannot be used in any earlier versions of ESLint. However, not allowing it also means having different schemas for user-defined and plugin configs, which is a significant downside.
+
+## Frequently Asked Questions
+
+### Why is `"#"` assigned as `null` in the example?
+
+This allows us to very quickly check `plugin.plugins["#"]` to see if there's a `"#"` that needs evaluating. The value actually doesn't matter because it will be replaced, so `null` seemed like a nice shorthand. Without the `"#"` entry in `plugins`, we'd need to search through `rules`, `processor`, `language`, and potentially other future keys to see if `"#"` was referenced. I considered that initially, but I think being explicit is a better idea.
+
+### Why is this functionality added to `FlatConfigArray` instead of `@eslint/config-array`
+
+In order to support named configs, we need the concept of a plugin. The generic `ConfigArray` class has no concept of plugins, which means the functionality needs to live in `FlatConfigArray` in some way. There may be an argument for supporting `extends` with just objects and arrays in `ConfigArray`, with `FlatConfigArray` overriding that to support named configs, but that would increase the complexity of implementation.
+
+If we end up not supporting named configs, then we can revisit this decision.
```

Comment:
As I mentioned in another comment, this makes it harder for eg. [`config-inspector`](https://github.com/eslint/config-inspector) to mimic the same logic since `FlatConfigArray`is currently only available internally in ESLint (see discussion in https://github.com/eslint/eslint/issues/18619)